array_proxy indexing

Hi guys,

With the new array_proxy class, I can't seem to be able to do things like this:

```
    array A = constant(0, 5);
    af_print(A.rows(0, 4).rows(0,3));
```

The first call to rows() gives an array_proxy, and the second call to rows() gives me a compiler error saying there is no rows() method in the array_proxy class:

```
main.cpp:52:27: error: no member named 'rows' in 'af::array::array_proxy'
```

I'm currently working around this by doing:

```
    array A = constant(0, 5);
    array B = A.rows(0, 4);
    af_print(B.rows(0,3));
```

Please could you advise if the functionality has been dropped or is an oversight?

--- a/include/af/array.h
+++ b/include/af/array.h
@@ -114,6 +114,23 @@ namespace af
             template<typename T> T scalar() const;
             template<typename T> T* device() const;
             void unlock() const;
+
+                  array::array_proxy row(int index);
+            const array::array_proxy row(int index) const;
+
+                  array::array_proxy rows(int first, int last);
+            const array::array_proxy rows(int first, int last) const;
+
+                  array::array_proxy col(int index);
+            const array::array_proxy col(int index) const;
+                  array::array_proxy cols(int first, int last);
+            const array::array_proxy cols(int first, int last) const;
+
+                  array::array_proxy slice(int index);
+            const array::array_proxy slice(int index) const;
+
+                  array::array_proxy slices(int first, int last);
+            const array::array_proxy slices(int first, int last) const;
         };
 
         //array(af_array in, const array *par, af_index_t seqs[4]);

--- a/src/api/cpp/array.cpp
+++ b/src/api/cpp/array.cpp
@@ -175,7 +175,7 @@ namespace af
 
 #define INSTANTIATE(T)                                                  \
     template<> AFAPI                                                    \
-    array::array(const dim4 &dims, const T *ptr, af::source src)       \
+    array::array(const dim4 &dims, const T *ptr, af::source src)        \
         : arr(0)                                                        \
     {                                                                   \
         initDataArray<T>(&arr, ptr, src, dims[0], dims[1], dims[2],     \
@@ -188,21 +188,21 @@ namespace af
         initDataArray<T>(&arr, ptr, src, d0);                           \
     }                                                                   \
     template<> AFAPI                                                    \
-    array::array(dim_t d0, dim_t d1, const T *ptr, af::source src)     \
+    array::array(dim_t d0, dim_t d1, const T *ptr, af::source src)      \
         : arr(0)                                                        \
     {                                                                   \
         initDataArray<T>(&arr, ptr, src, d0, d1);                       \
     }                                                                   \
     template<> AFAPI                                                    \
     array::array(dim_t d0, dim_t d1, dim_t d2, const T *ptr,            \
-                 af::source src)                                       \
+                 af::source src)                                        \
         : arr(0)                                                        \
     {                                                                   \
         initDataArray<T>(&arr, ptr, src, d0, d1, d2);                   \
     }                                                                   \
     template<> AFAPI                                                    \
     array::array(dim_t d0, dim_t d1, dim_t d2, dim_t d3, const T *ptr,  \
-                 af::source src) :                                     \
+                 af::source src) :                                      \
         arr(0)                                                          \
                                                                         \
     {                                                                   \
@@ -380,8 +380,7 @@ namespace af
 
     array::array_proxy array::row(int index)
     {
-        seq idx(index, index, 1);
-        return this->operator()(idx, span, span, span);
+        return const_cast<const array*>(this)->row(index);
     }
 
     const array::array_proxy array::col(int index) const
@@ -392,8 +391,7 @@ namespace af
 
     array::array_proxy array::col(int index)
     {
-        seq idx(index, index, 1);
-        return this->operator()(span, idx, span, span);
+        return const_cast<const array*>(this)->col(index);
     }
 
     const array::array_proxy array::slice(int index) const
@@ -404,8 +402,7 @@ namespace af
 
     array::array_proxy array::slice(int index)
     {
-        seq idx(index, index, 1);
-        return this->operator()(span, span, idx, span);
+        return const_cast<const array*>(this)->slice(index);
     }
 
     const array::array_proxy array::rows(int first, int last) const
@@ -416,8 +413,7 @@ namespace af
 
     array::array_proxy array::rows(int first, int last)
     {
-        seq idx(first, last, 1);
-        return this->operator()(idx, span, span, span);
+        return const_cast<const array*>(this)->rows(first, last);
     }
 
     const array::array_proxy array::cols(int first, int last) const
@@ -428,8 +424,7 @@ namespace af
 
     array::array_proxy array::cols(int first, int last)
     {
-        seq idx(first, last, 1);
-        return this->operator()(span, idx, span, span);
+        return const_cast<const array*>(this)->cols(first, last);
     }
 
     const array::array_proxy array::slices(int first, int last) const
@@ -440,8 +435,7 @@ namespace af
 
     array::array_proxy array::slices(int first, int last)
     {
-        seq idx(first, last, 1);
-        return this->operator()(span, span, idx, span);
+        return const_cast<const array*>(this)->slices(first, last);
     }
 
     const array array::as(af::dtype type) const
@@ -742,6 +736,24 @@ namespace af
 
         return array(arr);
     }
+
+    //FIXME: Check if this leaks
+#define MEM_INDEX(FUNC_SIG, USAGE)          \
+    array::array_proxy                      \
+    array::array_proxy::FUNC_SIG            \
+    {                                       \
+        array *out = new array(this->get());\
+        return out->USAGE;                  \
+    }
+
+    MEM_INDEX(row(int index)                , row(index));
+    MEM_INDEX(rows(int first, int last)     , rows(first, last));
+    MEM_INDEX(col(int index)                , col(index));
+    MEM_INDEX(cols(int first, int last)     , cols(first, last));
+    MEM_INDEX(slice(int index)              , slice(index));
+    MEM_INDEX(slices(int first, int last)   , slices(first, last));
+
+#undef MEM_INDEX
 
     ///////////////////////////////////////////////////////////////////////////
     // Operator =

--- a/test/index.cpp
+++ b/test/index.cpp
@@ -1099,6 +1099,38 @@ TYPED_TEST(IndexedMembers, MemFuncs)
 }
 
 
+TYPED_TEST(IndexedMembers, MemIndex)
+{
+    using namespace af;
+    array a = range(dim4(10, 10));
+    array b = a(seq(1,7), span);
+    array brow = b.row(5);
+    array brows = b.rows(5, 6);
+    array bcol = b.col(5);
+    array bcols = b.cols(5, 6);
+
+    array out_row = a(seq(1,7), span).row(5);
+    array out_rows = a(seq(1,7), span).rows(5, 6);
+    array out_col = a(seq(1,7), span).col(5);
+    array out_cols = a(seq(1,7), span).cols(5, 6);
+
+    ASSERT_EQ(0, where(brow != out_row).elements());
+    ASSERT_EQ(0, where(brows != out_rows).elements());
+    ASSERT_EQ(0, where(bcol != out_col).elements());
+    ASSERT_EQ(0, where(bcols != out_cols).elements());
+
+    array avol = range(dim4(10, 10, 10));
+    array bvol = avol(seq(1, 7), span, span);
+    array bslice = bvol.slice(5);
+    array bslices = bvol.slices(5, 6);
+
+    array out_slice = avol(seq(1,7), span, span).slice(5);
+    array out_slices = avol(seq(1,7), span, span).slices(5, 6);
+
+    ASSERT_EQ(0, where(bslice != out_slice).elements());
+    ASSERT_EQ(0, where(bslices != out_slices).elements());
+}
+
 TEST(Indexing, SNIPPET_indexing_first)
 {
     using namespace af;

